Lepton time-life info for NanoAOD

mbluj opened the pull request:
#### PR description:

This PR adds a customization to add lepton time-life related information to NanoAOD. The information contains:
* (refitted) secondary vertex for 3-prong taus (with covariance matrix),
* impact parameter vector (with significance) of the electron, muon and tau tracks wrt PV fitted with beam-spot constraint,
* parameters of electron, muon and tau tracks (with covariance matrix),
* primary vertex with beam-spot constraint (position and covariance matrix).

All those variables are used in the measurement of CP structure of the Higgs-tau Yukawa coupling and related analyses, e.g. measurement of polarization of Z in decays to taus.

Technically all lepton time-life variables are stored in a newly introduced data-format `TrackTimeLifeInfo` produced by a new producer `PATLeptonTimeLifeInfoProducer`  (a template class instantiated for `pat::Electron`, `pat::Muon` and `pat::Tau`) and then added by an instance of simple table producer. The PV with BS information is stored with a new instance (for `reco::Vertex`) of a simple table producer.
In the case of Run-2 input samples the PV with the BS constraint is produced with a custom producer (`PATRefitVertexProducer`) with miniAOD inputs (a new `reco::Vertex` collection with one item) as vertices with BS are not available in those samples.

In case a given new variable is not computed, e.g. SV for 1-prong taus or variables for a lepton is not passing selection, zeros are stored which are then effectively compressed in an output file.


#### PR validation:

Validated using 2022 and 2018 TTbar samples with customized default NanoAOD workflow obtained as follows:
```
cmsDriver.py --python_filename NanoAOD_lifeTimeCustom_cfg.py --eventcontent NANOAODSIM --customise_commands="process.add_(cms.Service('InitRootHandlers',EnableIMT = cms.untracked.bool(False)));process.MessageLogger.cerr.FwkReport.reportEvery=1000" --datatier NANOAODSIM --fileout file:NanoAOD_lifeTimeCustom.root --conditions auto:phase1_2018_realistic --step NANO:@LepTimeLife --filein file:some_ttbar2018file.root --era Run2_2018,run2_nanoAOD_106Xv2 --no_exec --mc -n -1
```
Added variables increase size of NanoAOD by ~8.5% (wrt 14_0_0_pre1 default) which is distributed between taus (~3%), electrons and muons (~2% each) and PV with BS (~1%)
In the current setup the variables are added to default workflow with `LepTimeLife` customization defining extended NANO step like this ` --step NANO:@LepTimeLife`.

Presentation at the XPOG meeting: [link](https://indico.cern.ch/event/1352781/#10-lepton-life-time-in-nano)

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

<!-- Please replace this text with any link to the master PR, or the intended backport release cycle numbers -->

It is foreseen to backport to 14_0_X.

cmsbuild commented:
A new Pull Request was created by  @mbluj  for master.

It involves the following packages:

- DataFormats/NanoAOD (**xpog**)
- PhysicsTools/NanoAOD (**xpog**)
- PhysicsTools/PatAlgos (**reconstruction, xpog**)


@cmsbuild, @vlimant, @jfernan2, @simonepigazzini, @mandrenguyen can you please review it and eventually sign? Thanks.
@emilbols, @seemasharmafnal, @andrzejnovak, @mbluj, @jdolen, @gkasieczka, @JyothsnaKomaragiri, @schoef, @nhanvtran, @jdamgov, @hatakeyamak, @ahinzmann, @azotz, @rovere, @gpetruc, @Senphy, @gouskos, @Ming-Yan, @mmarionncern, @missirol, @mariadalfonso, @AnnikaStein, @AlexDeMoor, @rappoccio, @demuller this is something you requested to watch as well.
@sextonkennedy, @rappoccio, @antoniovilela you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

vlimant commented:
enable nano

vlimant commented on PhysicsTools/PatAlgos/plugins/MiniAODRefitVertexProducer.cc:
```diff
@@ -0,0 +1,189 @@
+  produces<reco::VertexCollection>();
+}
+
+void MiniAODRefitVertexProducer::produce(edm::Event& evt, const edm::EventSetup& es) {
```
no need for MiniAOD in the module name : are you certain there does not exist one such module under RecoVertex ?

vlimant replied:
alright, following previous logic, I think you need to change it to `PATRefitVertexProducer`

mbluj replied:
OK

vlimant commented on PhysicsTools/NanoAOD/plugins/LeptonTimeLifeInfoTableProducer.cc:
```diff
@@ -0,0 +1,340 @@
+
+#include <cstring>
+
+namespace {
```
so as to avoid redefining this outside of the templated table producer, please split the producer in two. One that creates `LeptonTimeLifeInfo`, one that reads it back from edm to fill a table

mbluj replied:
It is in fact thins is what I wanted to avoid. The idea is to create a nanoAOD table directly without using intermediate products put to and then read from `edm::Event`. 
The `LeptonTimeLifeInfo` is meant to be an internal data structure rather than (potentially) persistent data-format, but to be able to use string parser to configure variables to be stored in the table the structure of the `LeptonTimeLifeInfo`  has to be parsed by reflex. So it is why is is defined outsite in DataFormats and added to classes.h/_def.xml. 
BTW, in the first version of this development intermediate collections of pat::Electron/Muon/Tau with set of `userFloat`s were produced, but I found it (i) not very flexible due to hard-coded names of the floats, (ii) not optimal due to need of producing quite big objects containing mainly duplicated information from original leptons. I can revert to it if it is preferred.

vlimant commented on DataFormats/NanoAOD/interface/LeptonTimeLifeInfo.h:
```diff
@@ -0,0 +1,73 @@
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+
+class LeptonTimeLifeInfo {
```
this class might belong elsewhere than NANO data formats, please investigate

mbluj replied:
Hmm, it is used only to produce nano tables and it is why I placed it here. It is a kind of analysis format, so maybe can go to PatCandidates or similar. But I have not strong feeling. Anyway, it is not expected to be a persistent data-format (as mentioned in other comment).

mmusich replied:
FWIW I do agree this looks better fitting somewhere else.

>  Anyway, it is not expected to be a persistent data-format (as mentioned in other comment).

the point is that I might want to be able to use this somewhere else in cmssw (e.g. in DQM or Alignment / Calibration) and don't want to drag in a dependence on `DataFormats/NanoAOD`.

mbluj replied:
I tend to agree. Do you have proposal where it will fit well?

mmusich replied:
>  Do you have proposal where it will fit well?

well, despite it's called `Lepton`-something I don't see anything super lepton-specific. I had in mind `DataFormats/VertexReco` (or `DataFormats/VertexReco`), perhaps after moving `Lepton` to `Track`.

mbluj replied:
OK, fine with me. 
I have not strong objections against `Lepton` -> `Track` however tracks usually have not SV, but of course a track can originate from a SV. Generally, if you have a particular use-case in mind the structure can be modified to better fit this use-case as currently it is tailored to what is needed in this particular development.

mmusich replied:
> you have a particular use-case in mind the structure can be modified to better fit this use-case.

the class layout looks good to me, thanks for creating this!

mbluj replied:
Moved to `DataFormats/VertexReco` and renamed changing `Lepton` to `Track`

vlimant commented on PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h:
```diff
@@ -452,7 +452,7 @@ class FirstObjectSimpleFlatTableProducer : public SimpleFlatTableProducerBase<T,
 
   static void fillDescriptions(edm::ConfigurationDescriptions &descriptions) {
-    edm::ParameterSetDescription desc = desc = SimpleFlatTableProducerBase<T, edm::View<T>>::baseDescriptions();
+    edm::ParameterSetDescription desc = SimpleFlatTableProducerBase<T, edm::View<T>>::baseDescriptions();
```
thanks for fixing this

vlimant commented:
hell @mbluj  are you able to present this at the upcoming xpog meeting (dec 13) ?

vlimant commented on PhysicsTools/NanoAOD/plugins/VertexTableProducer.cc:
```diff
@@ -74,6 +76,17 @@ class VertexTableProducer : public edm::stream::EDProducer<> {
   const std::string svDoc_;
   const double dlenMin_, dlenSigMin_;
+
+  typedef FuncVariable<reco::Vertex, StringObjectFunction<reco::Vertex>, int32_t> IntVar;
```
here again, I'd rather not have this kind of things redefined outside of the templated table producer, why did you have to do this ?

mbluj replied:
`FuncVariables` are defined inside `SimpleFlatTableProducerBase ` template. But, this producer is a custom table producer for vertices and to enable configurable optional parameters one needs to define instances of FuncVariable for `reco::Vertex` and various types (int32_t, float, etc.) somewhere. Is there better place to define those instances?

mbluj commented:
> hell @mbluj are you able to present this at the upcoming xpog meeting (dec 13) ?

I am participating at L1 workshop, but I can try to prepare a short presentation.

mbluj commented:
> > hell @mbluj are you able to present this at the upcoming xpog meeting (dec 13) ?
> 
> I am participating at L1 workshop, but I can try to prepare a short presentation.

I prepared a couple of slides, could you, please, give me a slot in agenda of the meeting?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43525/38289

- This PR adds an extra 36KB to repository

- Found files with invalid states:
  - DataFormats/NanoAOD/src/LeptonTimeLifeInfo.cc:
    - Added: dd9dc87657596a29ae1c6c45b1be7636709a053e
    - Deleted: 5a1ba08e08a880abb41e8b7c471b1b219b312fe5
  - DataFormats/NanoAOD/interface/LeptonTimeLifeInfo.h:
    - Added: dd9dc87657596a29ae1c6c45b1be7636709a053e
    - Deleted: 5a1ba08e08a880abb41e8b7c471b1b219b312fe5
  - PhysicsTools/PatAlgos/plugins/PATLeptonTimeLifeInfoUpdater.cc:
    - Added: 84637116006c50068010a54dfb208e128cb65285
    - Deleted: dd9dc87657596a29ae1c6c45b1be7636709a053e

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/BuildFile.xml  modified in PR(s): #43485
  - File  PhysicsTools/NanoAOD/plugins/VertexTableProducer.cc  modified in PR(s): #43487
  - File  PhysicsTools/NanoAOD/python/electrons_cff.py  modified in PR(s): #42607, #43430
  - File  PhysicsTools/NanoAOD/python/muons_cff.py  modified in PR(s): #43622
  - File  PhysicsTools/NanoAOD/python/vertices_cff.py  modified in PR(s): #43487

cmsbuild commented:
Pull request #43525 was updated. @cmsbuild, @hqucms, @jfernan2, @mandrenguyen, @vlimant can you please check and sign again.

mbluj commented:
The two last commits address review comments (hopefully):
* Auxillary TimeLifeInfo class moved from `DataFormats/NanoAOD` to `DataFormats/VertexReco` and renamed as suggested,
* Definition of variables dumped to tables decoupled from definition of simple table producers,
* Removed "heavy" use `#typedef` in Vertex and TimeLifeInfo table producers.

If the changes in this PR are accepted I will add newly added variables to NanoAOD DQM as this is still missing.

mbluj commented:
@vlimant, I wonder if you have had time to review update of this PR? Do you have any further comments?

vlimant commented on PhysicsTools/NanoAOD/plugins/VertexTableProducer.cc:
```diff
@@ -93,6 +101,35 @@ VertexTableProducer::VertexTableProducer(const edm::ParameterSet& params)
       dlenSigMin_(params.getParameter<double>("dlenSigMin"))
 
 {
+  if (params.existsAs<edm::ParameterSet>("optionalPvVariables")) {
```
please remove the `optionalPvVariables` (and DumpedVariables.h) that you are not using, and that can be implemented using `SimpleVertexFlatTableProducer` as you did, and should be done

mbluj replied:
The issue with using `SimpleVertexFlatTableProducer` in this context is that it is not possible to simply extend the "main" PV table produced by `VertexTableProducer.cc`. One can add a new table, even with the same name as the main PV one, but extension does not work. It is a scheduling problem. Maybe it is also because the main PV table is declared as singleton, while to produce table with 1 item from multi-item collection one should do the trick by setting max no. of inputs to 1 (it takes the leading input which is fine for PV). 
When the two PV tables (main and additional) have same name it is not visible a final user but although it works with NANOAOD output module I am not sure if it will work with NANOAODEDM output module which I have not tried.
Maybe I overlooked something - do you have any suggestions? And what is a problem with current implementation? It is too specific?

mbluj replied:
Any further comments, suggestions?

vlimant replied:
please follow the previous suggestion. if there's an issue with the extension, please fill a GH issue, or in the XPOG gitlab and we'll figure this out

mbluj replied:
OK.

vlimant commented on PhysicsTools/NanoAOD/plugins/LeptonTimeLifeInfoTableProducer.cc:
```diff
@@ -0,0 +1,334 @@
+#include <cstring>
+
+template <typename T>
+class LeptonTimeLifeInfoTableProducer : public edm::stream::EDProducer<> {
```
I think I did suggest to split the `LeptonTimeLifeInfo` producer and the table producer ; is there anything fundamental preventing this ?

mbluj replied:
I thought I convinced you during the meeting that it is better to produce directly the timeLifeInfo table rather than to produce collections of `LeptonTimeLifeInfo` put them to the Event and then get them back to produce the table. I think that the "split" approach would be justified only when there were other users of the `LeptonTimeLifeInfo` collection(s) than table producer or if there is a strong argument against custom table producers. I am not convinced that any of those is the case. But, I can miss something.
Anyway, it will not be a big deal to move to the "split" approach (which actually was used at initial stage of this development, but with pat leptons as "mediators"). If we decide to go this direction where the LeptonTimeLifeInfo producer should be placed, NanoAOD, PatAlgos other place?

mbluj replied:
Do you agree with my comment or you maintain your opinion about the "splitting"?

vlimant replied:
I understand that this could be complicated due to the track. Maybe keep a full object in `LeptonTimeLifeInfo` instead of a pointer, and make it non persistent

mbluj replied:
After more thinking: I am not sure if it is possible to ensure in the "split" approach that `LeptonTimeLifeInfo` items are sorted in the same way as corresponding leptons in their collections, This prevents extending simply lepton tables by adding time-life info columns... In other words this approach forces adding new time-life info tables which then can be linked with lepton ones by indexes. This is doable, but it is not what was originally intended. Is it what is requested?

vlimant replied:
you may use valuemap ?

mbluj replied:
@vlimant Could you please give more details about proposed solution with valuemap? Values maps for simple numerical types can be easily passed to extend tables of their "parent" collections with `externalVariables`, but there is not a way to do this with variables to be extracted from complex types/data-formats (at least I do not see how to). It can be probably implemented, but it is not straightforward as type of the complex data-format has to be passed and then phrased by a table filler.
Solution with producing a collection of time-life infos and then producing table on top of it with a simple-table-producer does not guarantee that there will be correct matching between infos and their parent leptons...
Maybe you propose to produce valuemap of indexes (`int`s) which can be used for matching between leptons and items of time-life info collection? Indexes then can be stored in lepton tables and can point to items in separate time-life info tables. This should work, however, it is not very elegant.

vlimant commented:
There are still a couple of technical issues, and to speed things up you can already add the DQM plots.
That being said, we would have to demonstrate a wide usage of the additional variables, in order to justify the significant event size increase

mbluj commented:
> There are still a couple of technical issues, and to speed things up you can already add the DQM plots. That being said, we would have to demonstrate a wide usage of the additional variables, in order to justify the significant event size increase

OK, I will add DQM plots.
The users is HIG HLepRare group and TAU POG for sure. I do not know if it is big enough community. If not, this development can enter as a customization similarly to other "special purpose" NanoAOD workflows (e.g BTV one).

vlimant commented on DataFormats/VertexReco/interface/TrackTimeLifeInfo.h:
```diff
@@ -0,0 +1,73 @@
+  void setTrack(const reco::Track* track) { track_ = track; }
+  const reco::Track* track() const { return track_; }
+  bool hasTrack() const { return track_ != nullptr; }
+  void setBFiled_z(float bField_z) { bField_z_ = bField_z; }
```
please fix the typo `setBFiled_z` to `setBField_x`

mbluj replied:
It is consistently `z`, do I miss something?

mmusich replied:
the method is called `setBFiled_z` (which looks like a type). I guess `setBField_z` was meant instead.

mbluj replied:
Thanks, now I got it!

vlimant commented on DataFormats/VertexReco/interface/TrackTimeLifeInfo.h:
```diff
@@ -0,0 +1,73 @@
+  GlobalPoint pca_;
+  GlobalError flight_cov_, pca_cov_, ip_cov_;
+  Measurement1D flightLength_, ipLength_;
+  const reco::Track* track_;
```
are `track_` and `bField_z_` really meant to be persisted ?

mbluj replied:
No.

vlimant commented:
Please go on with the technical changes requested to have it in the release.
We need to have this implemented as a customized nano, so please add them as such (i.e not in the central nano)

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43525/38528

- This PR adds an extra 164KB to repository

- Found files with invalid states:
  - DataFormats/NanoAOD/src/LeptonTimeLifeInfo.cc:
    - Added: dd9dc87657596a29ae1c6c45b1be7636709a053e
    - Deleted: 5a1ba08e08a880abb41e8b7c471b1b219b312fe5
  - DataFormats/NanoAOD/interface/LeptonTimeLifeInfo.h:
    - Added: dd9dc87657596a29ae1c6c45b1be7636709a053e
    - Deleted: 5a1ba08e08a880abb41e8b7c471b1b219b312fe5
  - PhysicsTools/PatAlgos/plugins/PATLeptonTimeLifeInfoUpdater.cc:
    - Added: 84637116006c50068010a54dfb208e128cb65285
    - Deleted: dd9dc87657596a29ae1c6c45b1be7636709a053e
  - PhysicsTools/PatAlgos/plugins/MiniAODRefitVertexProducer.cc:
    - Added: e16b1e31be7ed4c51ae9c64c1986068d6df4b161
    - Deleted: c3d0d11de6ae2c38b9ee0bcb6ef5ded47d1315cb

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h  modified in PR(s): #43734
  - File  PhysicsTools/NanoAOD/plugins/BuildFile.xml  modified in PR(s): #43485
  - File  PhysicsTools/NanoAOD/python/electrons_cff.py  modified in PR(s): #42607, #43734

vlimant commented:
having first the lepton lifetimeinfo in edm, then having a simpletableproducer for that type would allow to make an extension to the existing lepton tables.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43525/38712

ERROR: Unable to merge PR.

See log https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43525/38712/cms-checkout-topic.log

mbluj commented:
Time-life info moved from default workflow to customization function; it can be added with cmsDriver by `--step/-s NANO:@LepTimeLife`. It name of the customization fine? Another question: should DQM be also customized for workflow with time-life info? Currently new plots are added to the standard nano-DQM.
Last missing part is splitting producer of time-life and table producer.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43525/38716

- This PR adds an extra 64KB to repository

- Found files with invalid states:
  - DataFormats/NanoAOD/src/LeptonTimeLifeInfo.cc:
    - Added: bee552dbbbccdeda994105a8d395cb9b65027aa0
    - Deleted: 14deb3d677d0cde994b42f856825efe492c2972b
  - DataFormats/NanoAOD/interface/LeptonTimeLifeInfo.h:
    - Added: bee552dbbbccdeda994105a8d395cb9b65027aa0
    - Deleted: 14deb3d677d0cde994b42f856825efe492c2972b
  - PhysicsTools/PatAlgos/plugins/PATLeptonTimeLifeInfoUpdater.cc:
    - Added: 668216f57694df8c73071539a47d62b929cf0354
    - Deleted: bee552dbbbccdeda994105a8d395cb9b65027aa0
  - PhysicsTools/PatAlgos/plugins/MiniAODRefitVertexProducer.cc:
    - Added: 41220d3e3ee7d227eb090981cedbbd40e77b64fc
    - Deleted: ac66cea4d652ba447a0bb5ee161825825be34391

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/BuildFile.xml  modified in PR(s): #43485
  - File  PhysicsTools/NanoAOD/python/autoNANO.py  modified in PR(s): #43642

cmsbuild commented:
Pull request #43525 was updated. @vlimant, @mandrenguyen, @cmsbuild, @jfernan2, @hqucms can you please check and sign again.

vlimant commented:
Hi @mbluj , we came up with a solution for this particular problem and I'm going to make PR to add functionality to the nano producer to support your use case. Please hold on

mbluj commented:
@vlimant as #43916 is already merged I am rebasing this development on top of it. Therefore, I would like your advice in a few items:
* In this development `SimpleFlatTableProducer.h` was divided on `DumpedVariable.h` where variable classes are defined and `SimpleFlatTableProducer.h` where definition of actual tables producers are kept. Should I keep this division or revert to one common header file? Both have some pros and cons;
* In history of this development there are a few changes which were then reverted, e.g. updates in the vertex table producer. I tend to squash the history removing such back and forth changes. This way they will not pollute the history and make rebase simpler. Is it fine with you?
* What about backporting this to 14_0 (together with  #43916)? It will be good to have this in production release series. This backport will not affect any official workflows as new variables are added via customization. (This question can be addressed later).

mbluj committed:
Add time-life info structure

mbluj committed:
Module to refit PV with beam-spot constraint with miniAOD inputs

mbluj committed:
Simple fixes in SimpleFlatTableProducer

vlimant commented:
Hi, no need for `DumpedVariable.h` indeed, squashing things out where relevant would be great. We will see about the backport when we finalise this PR

mbluj committed:
Add time-life info through customization

mbluj committed:
NanoAOD DQM plots for lepton time-life variables

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43525/38850

- This PR adds an extra 60KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/python/autoNANO.py  modified in PR(s): #42607, #43642
  - File  PhysicsTools/NanoAOD/python/nano_cff.py  modified in PR(s): #42607, #43956

cmsbuild commented:
Pull request #43525 was updated. @mandrenguyen, @jfernan2, @vlimant, @hqucms, @cmsbuild can you please check and sign again.

mbluj commented:
@vlimant it was rebased and squashed. When squashing a lot of a (long) history of this development was lost, but I hope that it is not a problem. I think that now the important changes are clearly documented.
I am going to adjust also description of the PR to the current state.

mbluj commented:
@vlimant Is there still something missing in this PR? I would like to have it merged soon and then backport if to 14_0

vlimant commented on PhysicsTools/NanoAOD/python/taus_cff.py:
```diff
@@ -1,11 +1,13 @@
 import FWCore.ParameterSet.Config as cms
 from PhysicsTools.NanoAOD.common_cff import *
-from PhysicsTools.NanoAOD.nano_eras_cff import run3_nanoAOD_124
+from PhysicsTools.NanoAOD.nano_eras_cff import *
```
why ?

mbluj replied:
There is not any particular reason. I think I changed it for some test and then not reverted. It can be changed back, however I do not see problem with importing all eras. Should I revert it?

vlimant replied:
I prefer the import of only the things required

vlimant commented on PhysicsTools/NanoAOD/python/nanoDQM_cfi.py:
```diff
@@ -132,6 +132,32 @@
                 NoPlot('vidNestedWPBitmap'),
                 NoPlot('vidNestedWPBitmap_Fall17V2'),
                 NoPlot('vidNestedWPBitmapHEEP'),
+                Plot1D('ipLength', 'ipLength', 25, -0.25, 0.25, 'signed lenght of impact parameter (3d)'),
```
it seems a bit awkward to have the plot defined in mainstream configuration, please implement those in the customise, assuming that one would run the NANO and DQM step in the same process

mbluj replied:
OK.

vlimant commented on PhysicsTools/PatAlgos/plugins/PATRefitVertexProducer.cc:
```diff
@@ -0,0 +1,192 @@
+  desc.add<edm::InputTag>("srcLostTracks", edm::InputTag("lostTracks"));
+  desc.add<edm::InputTag>("srcEleKfTracks", edm::InputTag("lostTracks:eleTracks"));
+  desc.add<edm::InputTag>("srcBeamSpot", edm::InputTag("offlineBeamSpot"));
+  desc.add<bool>("useBeamSpot", true);
```
how about a `setComment` for the non trivial parameters

mbluj replied:
OK.

vlimant commented on PhysicsTools/NanoAOD/python/autoNANO.py:
```diff
@@ -35,5 +35,9 @@ def expandNanoMapping(seqList, mapping, key):
-                'customize': '@PHYS+@L1'}
+                'customize': '@PHYS+@L1'},
+    # Add lepton time-life info tables through customize combined with PHYS
+    'AddLepTimeLife' : {'customize': 'addTimeLifeInfo'},
```
no need for `AddLepTimeLife` entry, one can do `NANO:@omething+@LepTimeLife`. 
please use `PhysicsTools/NanoAOD/leptonTimeLifeInfo_common_cff.addTimeLifeInfo` and remove the import at top level

vlimant commented on PhysicsTools/NanoAOD/python/nano_cff.py:
```diff
@@ -28,6 +28,7 @@
 from PhysicsTools.NanoAOD.fsrPhotons_cff import *
 from PhysicsTools.NanoAOD.softActivity_cff import *
 from PhysicsTools.NanoAOD.l1trig_cff import *
+from PhysicsTools.NanoAOD.leptonTimeLifeInfo_common_cff import *
```
please remove the import from here, and use it only in autoNano

vlimant commented on PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h:
```diff
@@ -121,7 +121,7 @@ class ValueMapVariable : public ValueMapVariableBase<ObjType, TIn, ValType> {
                    bool skipNonExistingSrc = false)
       : ValueMapVariableBase<ObjType, TIn, ValType>(aname, cfg, std::move(cc), skipNonExistingSrc) {}
-  ValType eval(const edm::Handle<edm::ValueMap<TIn>> &vmap, const edm::Ptr<ObjType> &op) const {
+  ValType eval(const edm::Handle<edm::ValueMap<TIn>> &vmap, const edm::Ptr<ObjType> &op) const override {
```
was there anything failing without the `override` ?

mbluj replied:
No, but it clearly indicates that this overrides method from base class; it is also consistent with what is done for fill method.

mbluj committed:
Changes in time-life info customization

mbluj committed:
DQM plots for lepton time-life variables added as part of customization

mbluj commented:
@vlimant I addressed your comments in the recent commits. 
Some of them were small and simple changes (https://github.com/cms-sw/cmssw/pull/43525#discussion_r1495446294 and https://github.com/cms-sw/cmssw/pull/43525#discussion_r1495455795), so commits implementing them were squashed with previous commits to not pollute history. Commits updating customization and definition of DQM plots were kept for easier review, but I also plan to squash them afterwards.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43525/38967

- This PR adds an extra 32KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/python/autoNANO.py  modified in PR(s): #42607, #43642

cmsbuild commented:
Pull request #43525 was updated. @hqucms, @jfernan2, @cmsbuild, @vlimant, @mandrenguyen can you please check and sign again.

vlimant commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4f2664/37588/summary.html
**COMMIT**: bef9e1715774942f3d91e891cb34921d2e3d14a8
**CMSSW**: CMSSW_14_1_X_2024-02-20-1100/el8_amd64_gcc12
**Additional Tests**: NANO
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43525/37588/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 62 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-20-1100+4f2664/61231/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 44 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-20-1100+4f2664/61231/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3248389
 - DQMHistoTests: [Total failures: 3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-20-1100+4f2664/61231/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3248364
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-20-1100+4f2664/61231/triggerResults): no differences found


## NANO Comparison Summary



**Summary**:
 - You potentially [added 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNANO/CMSSW_14_1_X_2024-02-20-1100+4f2664/61230/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNANO/CMSSW_14_1_X_2024-02-20-1100+4f2664/61230/validateJR.html)
 - DQMHistoTests: Total files compared: 15
 - DQMHistoTests: Total histograms compared: 16430
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNANO/CMSSW_14_1_X_2024-02-20-1100+4f2664/61230/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 16430
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 14 files compared)
 - Checked 38 log files, 18 edm output root files, 15 DQM output files

**Nano size comparison Summary**:

| Sample | kb/ev | ref kb/ev | diff kb/ev | ev/s/thd | ref ev/s/thd | diff rate | mem/thd | ref mem/thd |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 2500.0 | 2.551 | 2.551 | 0.000 ( +0.0% ) | 5.23 | 5.30 | -1.2% | 2.237 | 2.221 |
| 2500.001 | 2.700 | 2.700 | 0.000 ( +0.0% ) | 4.65 | 4.76 | -2.3% | 2.686 | 2.652 |
| 2500.002 | 2.641 | 2.641 | 0.000 ( +0.0% ) | 4.87 | 4.91 | -0.8% | 2.676 | 2.655 |
| 2500.01 | 1.322 | 1.322 | 0.000 ( +0.0% ) | 9.57 | 9.83 | -2.6% | 2.366 | 2.291 |
| 2500.011 | 1.745 | 1.745 | 0.000 ( +0.0% ) | 5.23 | 5.21 | +0.5% | 2.547 | 2.468 |
| 2500.012 | 1.586 | 1.586 | 0.000 ( +0.0% ) | 7.59 | 7.52 | +0.8% | 2.466 | 2.430 |
| 2500.1 | 2.194 | 2.194 | 0.000 ( +0.0% ) | 5.32 | 5.32 | -0.0% | 2.087 | 2.070 |
| 2500.2 | 2.310 | 2.310 | 0.000 ( +0.0% ) | 6.07 | 6.00 | +1.2% | 2.000 | 1.983 |
| 2500.21 | 1.185 | 1.185 | 0.000 ( +0.0% ) | 4.31 | 4.40 | -2.1% | 2.294 | 2.260 |
| 2500.211 | 1.549 | 1.549 | 0.000 ( +0.0% ) | 3.74 | 3.82 | -1.9% | 2.382 | 2.340 |
| 2500.3 | 2.060 | 2.060 | 0.000 ( +0.0% ) | 12.71 | 12.68 | +0.2% | 1.988 | 1.966 |
| 2500.31 | 1.257 | 1.257 | 0.000 ( +0.0% ) | 20.13 | 20.02 | +0.6% | 2.381 | 2.357 |
| 2500.311 | 1.646 | 1.646 | 0.000 ( +0.0% ) | 12.18 | 12.82 | -5.0% | 2.461 | 2.409 |
| 2500.312 | 7.159 | 7.159 | 0.000 ( +0.0% ) | 1.39 | 1.46 | -4.8% | 1.703 | 1.690 |
| 2500.313 | 1.564 | 1.564 | 0.000 ( +0.0% ) | 6.78 | 6.69 | +1.3% | 1.056 | 1.042 |
| 2500.4 | 2.060 | 2.060 | 0.000 ( +0.0% ) | 12.68 | 12.75 | -0.6% | 1.985 | 1.967 |
| 2500.5 | 19.575 | 19.575 | 0.000 ( +0.0% ) | 1.36 | 1.27 | +6.7% | 1.372 | 1.354 |

mandrenguyen commented:
+reconstruction

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @rappoccio, @antoniovilela (and backports should be raised in the release meeting by the corresponding L2)

vlimant commented:
please prepare the backport PR asap

mbluj commented:
> please prepare the backport PR asap

OK. Can I backport the new typed value map table producer together with this development or two separate PR should be prepared? Or the producer is already backported?

vlimant commented:
backport both at the same time please

rappoccio commented:
+1

vlimant commented:
please work on a PR with a test workflow in the nano matrix for this feature ; following the example for MUO, EGM, etc (no need to have it in backport though)
